Pipeline design

toinehartman opened the pull request.

toinehartman committed:
Re-design framework using pipeline structure.

toinehartman committed:
Fixes & naming.

toinehartman committed:
Rewrite pico renaming for pipeline.

DavyLandman commented on src/main/rascal/examples/pico/Rename.rsc:
```diff
@@ -41,93 +41,46 @@ import util::FileSystem;
+    TModel tm = getTModel(cursor[-1]);
+    if (Tree t <- cursor
+      , tm.definitions[t.src]?) {
+            return {tm.definitions[t.src]};
```
shouldn't this be a set already?

toinehartman replied:
No, `tm.definitions` is a `map[loc, Define]`

DavyLandman commented on src/main/rascal/examples/pico/Rename.rsc:
```diff
@@ -41,93 +41,46 @@ import util::FileSystem;
-    //         solver.collectParseTree(m, renameByTree, rename(defRole, def, req.newName));
-    //     }
-    // }
+    r.msg(error("No definition for name under cursor", cursor[0].src));
```
we could also add overloads for `error`/`warning`/`info` next to this `msg`, just to help the user out.

possible include the format support we have in typepal.

DavyLandman replied:
and shouldn't this be an error message that the frameworkr generates automatically if you return an empty set (or throw an exception in this functiion)

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,176 @@ module refactor::Rename
-            println("-- Using cached tree for <l>");
+    void checkEdit(te:replace(loc range, _)) {
+        if (te in editsSeen) {
+            messages += error("Multiple replace edits for this location.", range);
```
what if the `loc` is the same, but the replacement isn't? maybe the `editSeen` should be a `set[loc]`?

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,176 @@ module refactor::Rename
+
+    map[str id, ChangeAnnotation annotation] annotations = ();
+    void registerAnnotation(str annotationId, ChangeAnnotation annotation) {
+        if (annotationId in annotations) throw "An annotation with id \'<annotationId>\' already exists!";
```
nit: should be an error message

toinehartman committed:
Rewrite modules renaming for pipeline.

toinehartman committed:
Improve edit checking and message registration.

toinehartman committed:
Short-circuit on errors.

toinehartman committed:
Move no-defs error to framework.

toinehartman committed:
Clean up warnings.

DavyLandman commented on src/main/rascal/examples/modules/Rename.rsc:
```diff
@@ -93,16 +91,13 @@ bool tryParse(type[&T <: Tree] tp, str s) {
-    if (/Tree t := modTree, "<t>" == d.id) {
-        renamer.collectTModel(modTree.src.top, renameByModel, rename(d, req));
+    set[str] names = {d.id | d <- defs};
+    if (/Tree t := modTree, "<t>" in names) {
```
normally you would pattern match on a more specific thing than all trees, for example `/Name n` or `/Identifier i`

toinehartman committed:
Introduce config/state ADT.

toinehartman committed:
Prevent pretty-printing all trees.

toinehartman committed:
Check all trees and find def/use candidates early.

toinehartman committed:
Implement renaming of overloads.

toinehartman committed:
Remove change annotations from framework.

toinehartman committed:
Move to typepal messages with formatting.

toinehartman committed:
Remove obsolete tests.

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,200 @@ module refactor::Rename
-      , bool reportCollectCycles = false
-      , bool debug = true
+      , TModel(Tree) tmodelForTree
+      , tuple[set[Define] defs, set[loc] uses](list[Tree] cursor, Tree(loc) getTree, TModel(Tree) getTModel, Renamer r) findCandidates
```
We have to think about the names of `findCandidates`.

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,200 @@ module refactor::Rename
+      , TModel(Tree) tmodelForTree
+      , tuple[set[Define] defs, set[loc] uses](list[Tree] cursor, Tree(loc) getTree, TModel(Tree) getTModel, Renamer r) findCandidates
+      , void(Define def, str newName, TModel tm, Renamer r) renameDef
+      , void(set[Define] defs, str newName, set[loc] candidates, TModel tm, Renamer r) renameUses
```
why here a set of defs and not the same as `renameDef`?

DavyLandman replied:
ah, reading a bit more, these are the definitions candidates found earlier?

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,200 @@ module refactor::Rename
+      , tuple[set[Define] defs, set[loc] uses](list[Tree] cursor, Tree(loc) getTree, TModel(Tree) getTModel, Renamer r) findCandidates
+      , void(Define def, str newName, TModel tm, Renamer r) renameDef
+      , void(set[Define] defs, str newName, set[loc] candidates, TModel tm, Renamer r) renameUses
+      , bool(set[Define] defs, Tree t, Renamer r) skipCandidate = bool(_, _, _) { return false; }
```
I think we don't need this, as the `findCandidates` can be the place where we filter (since we already open a tree just to determine if it should even be analyzed).

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,200 @@ module refactor::Rename
+        list[Tree] cursor
+      , str newName
+      , RenameConfig config
+      , bool debug = true) {
```
debug looks more like a parameter for the `RenameConfig`?

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,200 @@ module refactor::Rename
-      , void(str, value) writeStore = void(str _, value _) { throw "Not implemented"; }
-) = rsolver();
+      // Helpers
+      , void(str, loc) warning
```
shouldn't these interfaces be extended to match the `FailMessage` possibilities?

toinehartman replied:
I swapped the arguments and now use `value` for the location, but unfortunately, Rascal does not syntactically allow the use of varargs here.

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,200 @@ module refactor::Rename
 
-                fileModelTasks = modelTaskQueueCopy[f];
-                if (config.debug) println("<size(fileModelTasks)> tasks for model of <f>");
+    if (debug) println("Renaming <cursor[0].src> to \'<newName>\'");
```
make a local `debugPrint` function that does the if inside of it?

DavyLandman commented on src/main/rascal/refactor/TextEdits.rsc:
```diff
@@ -31,13 +31,8 @@ extend analysis::diff::edits::TextEdits;
 
-alias ChangeAnnotationId = str;
-
 data ChangeAnnotation
```
I'm not so sure we should have this as part of this framework, smells more like a thing we either have in VS Code, or try to bring to `TextEdits` module.

toinehartman replied:
Removed

DavyLandman commented on src/main/rascal/refactor/Rename.rsc:
```diff
@@ -29,200 +29,200 @@ module refactor::Rename
+            // If possible, merge with latest document edit
+            // TODO Just assign to docEdits[-1], once this issue has been solved:
+            // https://github.com/usethesource/rascal/issues/2123
+            docEdits = [*pre, changed(f, prev + e)];
```
better to say: `c:changed(f, _)` and then `c[changes=f.changes + e]` such that externally defined keywords of `c` are not lost due to this merge.

toinehartman committed:
Rewrite framework based on latest design insights

- Change staging to reduce amount of iterations over project files
- Provide sensible, generic default/fallback stage implementations
- Use extension mechanism to override defaults when needed

toinehartman committed:
Move debug flag to config.

toinehartman committed:
Align message helpers with FailMessage.

toinehartman committed:
Preserve change annotations on text edits.

toinehartman committed:
Do not rely on default implementation to throw error.
